RFC: Type token theming

mattrosno opened the pull request:
As an off-shoot of https://github.com/carbon-design-system/carbon/pull/3502, an effort needs to take place to determine strategic theming and white-labeling capabilities. Until then, there's a need to use Carbon in editorial pages. This RFC shows how this can be achieved by adding a minimal number of component tokens. This RFC does not address theming strategy. Changes to the `@carbon/type` are for demonstration purposes; may be cleaner ways to author those mixins/functions when doing this for real.

**Note:** the change to `_container.scss` is to emulate a product's Sass build.

# Accordion design

http://react.carbondesignsystem.com/?path=/story/accordion--default

![image](https://user-images.githubusercontent.com/1691245/61974066-409fe580-afab-11e9-9f6e-c9743e592627.png)

The default accordion has 14px font size for both header and content. The accordion item height computes to 32px via 6px top/bottom padding. The icon is 16px.

https://deploy-preview-3554--carbon-components-react.netlify.com/?path=/story/accordion--default

![image](https://user-images.githubusercontent.com/1691245/61974118-65945880-afab-11e9-9f25-7b5e88ccebb5.png)

The editorial theme intends for the accordion item to be 48px tall with 16px font size and a 20px icon. The changes to item height and icon size can't be achieved by setting relative to font size.

# Accordion new tokens

Only two new tokens are needed.

`$accordion-heading-height` defaults to `$carbon--spacing-07` which is 32px. This is the intended computed height that we can use to set properties to achieve that. The top/bottom in the accordion styling is refactored to use that new token instead of hardcoding 6px. That means we can set the new token to 48px and the accordion styling will compute the new top/bottom heading padding.

`$accordion-type-tokens` defaults to `$tokens`. ([$tokens Sassdoc](https://github.com/carbon-design-system/carbon/blob/master/packages/type/docs/sass.md#tokens-variable)) That means when doing your Sass build you can override a type token from 14px to 16px for that component.

# Icon sizing

To avoid adding new component tokens to address icon sizing, we know when an icon should follow font size in components. Per the [iconography](https://www.carbondesignsystem.com/guidelines/iconography/usage) guidelines, we know that 16px icons are meant to be pared with 14px font size, 20px icons with 16px font size. That means we can create a Sass map for translation.

```scss
$carbon--type-icon-scale: (
  0.875rem: rem(16px),
  1rem: rem(20px),
);
```

With the help of a Sass function, we can then use this private-scoped Sass variable to get icon size based on it's paired font-size.

```scss
$accordion-arrow-size: carbon--type-icon-size(
  map-get(map-get($accordion-type-tokens, 'body-long-01'), 'font-size')
);
```

This won't need done in every component, but if an icon is inline, it makes sense.

# Sass build

This provides many options when building the accordion styling.

## 1) Default

```scss
@import '~carbon-components/src/components/accordion/accordion';
```

No change to the default accordion.

## 2) Override default

```scss
$accordion-type-tokens: map-merge(
  $tokens,
  (
    'body-long-01': $body-long-02,
  )
);
$accordion-heading-height: $carbon--spacing-09;
@import '~carbon-components/src/components/accordion/accordion';
```

This tells the accordion to use `body-long-02` (16px font size and related properties) in all uses of `body-long-01` in the accordion. Also, that the accordion item should compute to 48px tall.

## 3) Zoned

What if you want to use the default and the override on the same page?

```scss
@import '~carbon-components/src/components/accordion/accordion';

.editorial {
  $accordion-type-tokens: map-merge(
    $tokens,
    (
      'body-long-01': $body-long-02,
    )
  ) !global;
  $accordion-heading-height: $carbon--spacing-09 !global;
  @include accordion;
}
```

This first builds the default accordion. Then in a `editorial` namespace class builds an accordion with the editorial theme. This would be very useful so the UI shell elements (site header, left/right column, etc) could remain default theme with the editorial theme only getting applied to the page's main content.

# Considerations

## 1) Inconsistent type changes in a component

This theming approach overrides all instances of a type token in a given component. It wouldn't support 16px font size in the accordion item header and 14px font size in the accordion content. One could argue that's an acceptable constraint.

## 2) Breaking changes

```scss
.#{$prefix}--accordion__title {
    @include type-style('body-long-01', $accordion-type-tokens);
}
```

Since the custom build is overriding `$accordion-type-tokens`, if `body-long-01` is updated to `body-short-01` in the future because that may actually be a better token for the accordion title, is that a breaking change because somebody may be theming expecting `body-long-01` to change accordion styles?

mattrosno committed:
refactor(accordion): build-based accordion theming

mattrosno committed:
refactor(accordion): remove unused mixin

joshblack commented on packages/components/src/components/accordion/_accordion.scss:
```diff
@@ -15,6 +15,19 @@
+  /// Accordion arrow size in rems
+  /// @access private
+  /// @group accordion
+  $accordion-arrow-size: carbon--type-icon-size(
```
Do we need to clamp this value to one of: `[16, 20, 24, 32]`?

mattrosno replied:
Yes. I talked with @shixiedesign to confirm these font-size and icon size pairings:

```
12 16
14 16
16 20
20 24
24 32
```

shixiedesign replied:
Maybe we can omit the 12 16 pairing all together? It's not ideal (would want to shrink icon more and we don't have that size). We already have type size mapping of all 4 icon sizes we support without mentioning the 12px type.

joshblack commented on packages/components/src/components/accordion/_accordion.scss:
```diff
@@ -15,6 +15,19 @@
+  /// @type Map
+  /// @access public
+  /// @group accordion
+  $accordion-type-tokens: $tokens !default !global;
```
Is this so that we can do the lookup by string? Is there anyway we could pass the variable directly instead or is something preventing that?

mattrosno replied:
This just so you could isolate accordion type token overrides. E.g. in your build:

```scss
$accordion-type-tokens: map-merge(
  $tokens,
  (
    'body-long-01': $body-long-02,
  )
);
$accordion-heading-height: $carbon--spacing-09;
@import '~carbon-components/src/components/accordion/accordion';

// button's `$tokens` unaffected by accordion overrides
@import '~carbon-components/src/components/button/button';
```

---

The other approach that would eliminate the need of a new component type token would be:

* Every component has dedicated `_mixins.scss`
* So you could import those mixins without outputting CSS
* Then you could:

```scss
@import '~carbon-components/src/components/accordion/mixins';

// override accordion's use of `body-long-01` so font-sizes and related properties are 16px instead of 14px
$my-tokens: map-merge(
  $tokens,
  (
    'body-long-01': $body-long-02,
  )
);

@include accordion($my-tokens);
```

* The mixin's token parameter would default to `$tokens` so the parameter would be optional
* Maybe the mixin parameter could be of type `argslist` so you could pass any override? Don't know.

mattrosno replied:
Recent commit removes the need for exposing a new public `$accordion-type-tokens` variable.

joshblack commented on packages/components/src/globals/scss/_theme-tokens.scss:
```diff
@@ -158,6 +158,12 @@
+  /// @type Value
+  /// @access public
+  /// @group accordion
+  $accordion-heading-height: $carbon--spacing-07 !default !global;
```
Did we confirm with design if these things are related at all? Would be interesting if we could use our spacing tokens in this way or not, mostly because I think some spacing tokens don't hit the pixel grid.

mattrosno replied:
After reading the guidelines, I think we can be more clear how spacing tokens should be used within a component, and layout tokens should be used outside components as you assemble pages. That's my understanding. I'll open a separate PR to confirm.

photodow replied:
While I was working on the technical audit I leaned towards using spacing tokens, but after the fact, I also wondered if it made more sense to use the mini-unit function instead?

```scss
$accordion-heading-height: miniUnits(4) !default !global;
```

shixiedesign replied:
We discussed this in the morning and personally, I don't think spacing tokens are *intended* for this use (which is strictly speaking a height), but because it still influences the component, I can see an argument using it. I like the option of `miniUnits(4)`, more clear it's 32px at a glance.

joshblack replied:
@shixiedesign is there anything design-related (or token related) that this decision would be based off of that we could parameterize? The decision in this case being the height of the accordion header

shixiedesign replied:
I think we don't tokenize a component height because we expect line wrap during translations. This is why we center text in a component using `padding-top` and `padding-bottom`. When `padding-top` and `padding-bottom` are used to center an element, it's often random/odd values like 6px or 3px. It make sense to have a function to calculate it from a desired final height. But we should make sure the height is not a hard coded value to ensure translation & line wrap is not broken. FYI @mattrosno 

To answer your question, @joshblack we try to keep our component heights to multiples of eight (assuming no line wrap). We could potentially create parameters for heights, maybe reference data table's? Below are heights used in data table row, which are also common height values for a lot of our components (button, accordion, inputs, etc). We'd need to add 40px, though!

![image](https://user-images.githubusercontent.com/15144993/62079917-4436a500-b215-11e9-81a3-208eb767be5e.png)

joshblack replied:
@shixiedesign I was moreso curious what factors would design use that influence the height of a panel, and maybe we could use that to determine what it should be for a given font face / font size / line-height pairing.

For example, the padding has been updated to derive from the type token's line-height and subtracting that from the accordion heading height.

Was curious if there was some kind of design decision we could embody in code that says if your token is this font size / line height, then your height should be ___.

mattrosno replied:
@shixiedesign `$accordion-heading-height` doesn't tokenize the `height` CSS property in this PR. It's used to set top and button paddings. The end result is no change for it's current styling; everything will still wrap as expected.

mattrosno replied:
I'll use `carbon--mini-units(4)` instead of `$carbon--spacing-07` and move this discussion to a separate PR after evaluating our current use of spacing tokens across all components.

Our components are very inconsistent on when we use mini unit functions, or spacing tokens, or layout tokens.

shixiedesign replied:
@joshblack I see. I think there isn't 1-1 relationship between typography and height of a component. Height of a component is much more dependent on its function, so unfortunately it might not be easily computed. It's safe to say there are min height requirements. Eg. 14px type –> min height of a box containing that type should be 24px. 16px type –> min height should be 32px. but what we end up picking as the height of the component is usually based on things like how visible it should be? Eg. button should be more visible / bigger than accordion. What's usually around this component? Eg. button height should match input field if they line up. Does that answer your question?

joshblack commented on packages/components/src/globals/scss/_typography.scss:
```diff
@@ -45,6 +45,6 @@ $base-font-size: 16px !default;
 /// @group global-typography
-@mixin type-style($name, $fluid: false) {
-  @include carbon--type-style($name, $fluid);
+@mixin type-style($name, $tokenOverrides: $tokens) {
```
What do we need to pass `tokenOverrides` in for? Is it unsafe for us to assume that the token we're trying to include isn't available as the current value?

mattrosno replied:
I need to revisit this. I see I removed `$fluid` as a param just to get the concept working.

mattrosno replied:
Recent commit reverts this change.

joshblack commented on packages/components/src/globals/scss/vendor/@carbon/elements/scss/type/_styles.scss:
```diff
@@ -671,3 +672,21 @@ $tokens: (
+/// @return {Number} Icon size in rems
+/// @access public
+/// @group @carbon/type
+@function carbon--type-icon-size($font-size) {
```
The intent behind these additions seem great, should definitely bring them upstream. Are there any other additions that'd make working with stuff in this PR easier?

photodow replied:
:+1: I'm loving the idea here as well. While this is the general rule we're following, there are a couple of components like the code snippet family where the icon doesn't change its size to `20px`.

joshblack replied:
@photodow I think the rule we want to codify is that the icon size matches the type size, and is contained with the set `[16, 20, 24, 32]`

mattrosno replied:
My thinking was only components that had a clear type/icon pairing (e.g. if they are inline) would use the new `carbon--type-icon-size()`. If a component uses a standalone icon (no inline text), then the component shouldn't use this function.

asudoh commented on packages/react/.storybook/_container.scss:
```diff
@@ -1,9 +1,88 @@
-// as `src` directory is _not_ meant to be shipped in our NPM package.
-// Use e.g. `@import '~carbon-components/scss/globals/scss/styles.scss'` instead.
-@import '~carbon-components/src/globals/scss/styles.scss'; // SEE THE NOTE ABOVE
+@import '~carbon-components/src/globals/scss/feature-flags';
```
Instead of maintaining the component-level token usage pattern here, would we want to do the two?

- Writing some tests
- Documentation

mattrosno replied:
I don't follow. Changes to `_container.scss` are just for demonstration purposes only as specified in the RFC description. I don't intend on this PR getting merged with any `_container.scss` changes.

asudoh replied:
Your RFC description said it's for demonration, but didn't know that it's not meant for merge - Thank you for clarifying @mattrosno!

photodow commented on packages/components/src/components/accordion/_accordion.scss:
```diff
@@ -39,7 +52,16 @@
+                'line-height'
+              )
+          ) / 2
+      )
```
:+1: Loving this

photodow commented on packages/react/.storybook/_container.scss:
```diff
@@ -1,9 +1,88 @@
+$accordion-type-tokens: map-merge(
+  $tokens,
+  (
+    'body-long-01': $body-long-02,
```
I'm really struggling with overriding `body-long-01`. As I mentioned last week when you were discussing this option with me, from my point of view the intent behind these tokens changes when we override it. As mentioned in the pull request this wouldn't allow there to be a 14px and 16px within the same component. You also now have two different tokens set to the same exact value.

I fear the bigger issue here is how we communicate this to our adopters and contributors. It sounds like our whole design kit will have to change. The inconsistency is also wild where `$body-long-01` for one component doesn't look the same for `$body-long-01` in another component. When someone says `$body-long-01` which value are we talking about now?

joshblack replied:
Similar to color, we want to make sure that the tokens are being applied consistently across a global theme or across a zone. The ideal here is that `body-long-01` corresponds to specific moments of type usage, similar to how `interactive-01` is used. Here's the table on our site that shows some role examples for color tokens:

![image](https://user-images.githubusercontent.com/3901764/62075454-b655bc80-b20a-11e9-835b-4aa2af38a89a.png)


In the case of changing things within a component, this is where the zone theming idea comes in. We change the values of tokens for a given component context, but not for others. In this case, that would mean that these tokens get updated to their editorial theme equivalents in components that need to be in an editorial zone (like accordion).

If we look at tokens for the values they represent, then it's similar to just inlining the value directly. Instead, we should choose tokens for the design decisions they represent. 

Just a quick aside, this is definitely our first pass at tokens for type. If they feel off, we can definitely look into how we are applying these tokens and potentially updating them in places where their usage is not consistent.

photodow replied:
@joshblack I'm struggling to see how replacing `$interactive-02` is similar to replacing `$body-long-01`, and to be fair – I've told Matt this too – I could be way off on my understanding of type tokens.

It's my understanding that the type tokens are more elemental in nature, and ultimately part of the IBM Design Language. I saw it as almost a 1:1 tokenization of the IDL type sets. Similar to how you've tokenized the IDL colors. These are such elemental values that you don't want to change, right? It's how we communicate IBM's point of view of visual styling with one another. I look at setting `$body-long-01` to `$body-long-02` as changing its intent, and erasing a part of the IDL, and watering down how we communicate with each other.

With your color theming work you're not changing `$ibm-color__blue-50` to `$ibm-color__red-60`. It looks more like you've extracted that step out and defined specific meaning to those tokens like `$ui-background` is for UI Background and assigned IDL color value to it. If `$interactive-01` was set to `$ui-background` that would muddy the water. Just because their value is the same now doesn't mean their intent is, and thus that intent could direct a different value down the road.

If you want every component to use the same default type-token I don't personally see how `$body-long-01` takes on that intent. From my perspective, there should be a token like `$type-style-default` that extracts it a level and adds the intent y'all are discussing similarly to how the color theming works. That way your default might be set to `$body-long-01` out of the box, but dotcom it might be set to `$body-long-02`. Both `$body-long-01` and `$body-long-02` would maintain its original IDL values.

I hope that makes sense. Just something to think about I guess, so take it for what it is. 🤷‍♂️

joshblack replied:
> It's my understanding that the type tokens are more elemental in nature, and ultimately part of the IBM Design Language. I saw it as almost a 1:1 tokenization of the IDL type sets.

Ah, this could definitely be where the problem is. Similar to colors, type tokens also have roles: https://www.carbondesignsystem.com/guidelines/typography/productive as I said earlier, if one is being used in error we can definitely fix it 👍 this is our first attempt at this.

From the IDL standpoint, we're definitely blurring things here as the two teams are getting closer and closer. The IDL will define our type face, IBM Plex, along with things like:

- Type scale
- Font weights
- How to pair font sizes with line heights
- How to pair font sizes with letter spacing

But when it comes to the design system, our intent is not to create a set of type tokens and map them to color variables like we have for `carbon--blue-50`. Things like the type scale, font weights, etc. do have corresponding variables that would mirror these parts of type. However, as we approach tokens, these do have role-based usage exactly like our color tokens.

Put another way:

Variables corresponding to the IDL:
- `carbon--blue-60` for color
- `carbon--type-scale` for type scale
- `carbon--font-weights` for font weights
- `carbon--font-families` for font families

Tokens corresponding to how we combine the above variables in the system:
- `interactive-01`: uses blue-60 under the hood, applied to interactive elements
- `body-long-01`: uses a combination of type scale, weight, font family, letter spacing, and line height under the hood. Used for 4 lines of text or more. 

It's important that things like `body-long-01` are **not** coming from the IDL, otherwise there would be no way for teams outside of IBM to theme the type styles of carbon with a font face different than IBM Plex

photodow replied:
@joshblack thanks for the clarification!

If that's true I fear there is a disconnect somewhere within the process chain. If that disconnect is a misunderstanding on our team's part, then I fear we might need to go back to design since it was designed on a false pretense, and that this RFC might be trying to retrofit a solution based on a design using the wrong methodology and technical limitations. Because if what you're saying is true, dotcom should probably consider providing its own typeset, and reassign specific tokens like `$body-long-01` at a global level which might change how we spec out the work, change how we build the design kit, and even how we solve this expressive theming.

I've already reached out to @oliviaflory, and we've been working to realign on this point to figure out next steps here.

joshblack replied:
Sounds good! Sorry about the confusion here @photodow, let me know what we could write about/document that would help prevent this in the future! 🙏

mattrosno committed:
refactor(accordion): new carbon--type-theme mixin

- only keep the good parts of previous commit

mattrosno commented:
# Accordion

I have updated the accordion example so the PR uses one new component token (`accordion-heading-height`) and accordion type token can be set with a new `carbon--type-theme` mixin.

Editorial theme example (via Sass build).

```scss
$accordion-type-tokens: map-merge(
  $tokens,
  (
    'body-long-01': $body-long-02,
  )
);
$accordion-heading-height: carbon--mini-units(6);
@include carbon--type-theme($accordion-type-tokens) {
  @import '~carbon-components/src/components/accordion/accordion';
}
```

mattrosno committed:
refactor(button): allow type overrides

mattrosno commented:
# Button

This RFC isn't intended for merge. To confirm this approach of building an editorial theme, I refactored the button so you'd only need to do this to have buttons with 16px font-size and 20px icon size while keeping the default height at 48px.

https://deploy-preview-3554--carbon-components-react.netlify.com/?path=/story/buttons--default

```scss
$button-type-tokens: map-merge(
  $tokens,
  (
    'body-short-01': $body-short-02,
  )
);
@include carbon--type-theme($button-type-tokens) {
  @import '~carbon-components/src/components/button/button';
}
```

In doing so, I removed the `height` and `min-height` properties and fixed a bug so default icon-only button is 48px square. I simplified tokens by being aggressive with deprecations (because this is a RFC).

Note: our icon-only buttons at field and small sizes are not square. As-is those modifiers adjust height not width. I kept that functionality as making those square would be a separate effort.

shixiedesign commented:
These look good to me! 🎉 
![image](https://user-images.githubusercontent.com/15144993/62166831-600a7b80-b2e7-11e9-95ee-e90b36c45083.png)

asudoh commented on packages/react/.storybook/_container.scss:
```diff
@@ -1,9 +1,87 @@
+);
+$accordion-heading-height: carbon--mini-units(6);
+@include carbon--type-theme($accordion-type-tokens) {
+  @import '~carbon-components/src/components/accordion/accordion';
```
As I mentioned offline, the approach of `@import` of our component Sass file as `@content` conflicts with our `import-once` approach. Probably `@include accordion` instead by making it public?

mattrosno replied:
@asudoh that works here. We'd just need to make a pass making sure that every component has a public entrypoint mixin. Most (not all) have entrypoint mixins, and pretty sure they are all marked as private right now. E.g. https://github.com/carbon-design-system/carbon/blob/master/packages/components/docs/sass.md#accordion-mixin

asudoh replied:
BTW I meant:

```scss
@import '~carbon-components/src/components/accordion/accordion';

...

@include carbon--type-theme($accordion-type-tokens) {
  @include accordion;
}
```

(No `@include exports('button')` is needed)

mattrosno replied:
@asudoh that would emit the accordion styles twice. I think we're good here with how I have this working with `import-once`. I'll mention this in the recommended approach going forward.

asudoh replied:
@mattrosno Do you mean that the use case is having accordion style _only_ with custom theme?

mattrosno replied:
Correct.

mattrosno committed:
refactor(button): add import once

mattrosno commented:
# Import Once & Sass Partials

The latest commit shows how we can standardize component Sass so each component has an entrypoint Sass file, `_mixins.scss`, and `_variables.scss`. This would put us in the right direction for component packages, and would allow custom builds to import just the mixins, then using the `import-once` package decide when to import each component.

@asudoh does this address your question about `import-once`?

Accordion and button example:

```scss
$accordion-type-tokens: map-merge(
  $tokens,
  (
    'body-long-01': $body-long-02,
  )
);
$accordion-heading-height: carbon--mini-units(6);
@import '~carbon-components/src/components/accordion/mixins';
@include carbon--type-theme($accordion-type-tokens) {
  @include exports('accordion') {
    @include accordion;
  }
}

$button-type-tokens: map-merge(
  $tokens,
  (
    'body-short-01': $body-short-02,
  )
);
$button-field-height: 3rem;
@import '~carbon-components/src/components/button/mixins';
@include carbon--type-theme($button-type-tokens) {
  @include exports('button') {
    @include button;
  }
}
```

mattrosno commented on packages/components/src/globals/scss/vendor/@carbon/type/scss/_styles.scss:
```diff
@@ -468,6 +468,44 @@ $display-04: (
+/// @type Map
+/// @access private
+/// @group @carbon/type
+$defaultTokens: (
```
@joshblack to my knowledge there's no way to clone a Sass variable. This needs confirmed in a separate PR, but we may need to create two token maps: one that we override, and one used to determine when to reset the overrides.

That simplifies the theming mixin to:

```scss
@mixin carbon--type-theme($themedTokens: $tokens) {
  $tokens: $themedTokens !global;

   @content;

   // Reset to default type tokens after apply in content
  @if $tokens != $defaultTokens {
    @include carbon--type-theme($defaultTokens);
  }
}
```

asudoh replied:
@mattrosno Just wondered - Is cloning a Sass map what you wanted to do? Just in case if so, it can be done by map-walk technique.

mattrosno replied:
@asudoh I'm not familiar with "map-walk" technique to clone a Sass map. Can you share an example?

asudoh replied:
Actually for simple cloning we don't seem to do map-walking - We can simply do: `map-merge((), $source-map)`

mattrosno committed:
refactor(text-input): quick pass at 16px theme

- standardize sass partials
- theme type, input height, icon size

photodow commented on packages/components/src/globals/scss/vendor/@carbon/elements/scss/type/_inlined/_styles.scss:
```diff
@@ -468,6 +468,44 @@ $display-04: (
   ),
 ) !default;
 
+/// Used to compare with `$tokens` map so tokens can be reset to their default
```
@mattrosno out of curiosity, why is this file duplicated 4 times in this PR?

> packages/components/src/globals/scss/vendor/@carbon/type/scss/_styles.scss 
> packages/components/src/globals/scss/vendor/@carbon/type/scss/_inlined/_styles.scss
> packages/components/src/globals/scss/vendor/@carbon/elements/scss/type/_styles.scss
> packages/components/src/globals/scss/vendor/@carbon/elements/scss/type/_inlined/_styles.scss

mattrosno replied:
We inline elements packages for teams who can't configure their Sass loader to import directly from `node_modules`. Those duplicated files are generated in our build. In this example, I'm making edits to the `_styles.scss` from the `@carbon/type` package and the build does the rest.

photodow commented on packages/components/src/components/button/_mixins.scss:
```diff
@@ -5,7 +5,242 @@
+      $hover-tertiary,
+      currentColor,
+      $active-tertiary,
+      1px
```
Because the tertiary button changes its border width here, the tertiary button is `44px` instead of `48px`. We might need to consider adjusting something about the tertiary to make this work.

<img src="https://user-images.githubusercontent.com/3793636/62340518-78c09000-b4a5-11e9-9f25-ebf1092f9de4.png" width="267px" />

shixiedesign replied:
😱

mattrosno replied:
@shixiedesign Are tertiary buttons supported? I don't see it in the React storybook... assuming they are still supported we can handle tertiary buttons with this now that we're no longer hardcoding `height`:

```scss
.#{$prefix}--btn--tertiary {
  @include button-theme(...);

  $button-padding-y: (
      $button-height - rem(1px * 2) -
        map-get(map-get($tokens, 'body-short-01'), 'line-height')
    ) / 2;

  padding-top: $button-padding-y;
  padding-bottom: $button-padding-y;

  // rest
}
```

photodow commented on packages/components/src/components/button/_mixins.scss:
```diff
@@ -14,13 +249,32 @@
   @include reset;
   @include type-style('body-short-01');
```
Because `btn-sm` leverages `button-base` without overriding the `type-style`, `btn-sm` also takes on the default type style. It should maintain at the original `body-short-01` size instead of resizing with the rest of the buttons.

<img src="https://user-images.githubusercontent.com/3793636/62381310-d0e0ac00-b510-11e9-8021-8663799853fe.png" width="317" />

It should be more like this...

<img src="https://user-images.githubusercontent.com/3793636/62381436-1a30fb80-b511-11e9-8be3-d086de02e45e.png" width="317" />

mattrosno replied:
From @oliviaflory in group DM:

> If small button at 14px is causing a problem, I think we could live without it?

This RFC is attempting to modify type tokens on a **per-component** basis (not **per-variant/modifier/state** basis) to see if that is good enough, to prevent introducing new component tokens to address this specific use case.

Is this an isolated case where a component modifier doesn't get the type token override? Or is this a pattern that we'll see across many components? This seems like an outlier so far.

photodow replied:
So I could be wrong, and this could totally be out of scope for this RFC, but I was looking at it as *variants/modifiers/states* are part of the component, and if we are testing out the validity of a per-component basis it should involve this update and proof out if it can handle outliers like these. I'll leave it for you to decide, I just wanted to point it out for consideration just in case, and so we don't overlook these things as we continue to push forward. 😄

photodow replied:
@mattrosno if that spec changed I must have missed it. @oliviaflory can you chime in on the `14px`/`16px`?

oliviaflory replied:
@mattrosno @photodow ah yes, I *did* say we could live with 16px in our group chat. I think that makes sense from a system perspective. We had been considering leaving small button at 14px for instances within other components, like `code snippet` or `data table`, but I do think 16px looks alright there.

You might hate me for this, but could I retract my earlier statement on changing medium button to 48px? If we keep small button at 32px, I think it makes more sense to leave medium at 40px so that we are not providing two "different" variations that look the same visually.

photodow replied:
> You might hate me for this, but could I retract my earlier statement on changing medium button to 48px? If we keep small button at 32px, I think it makes more sense to leave medium at 40px so that we are not providing two "different" variations that look the same visually.

@oliviaflory even though the field/medium at `40px` button would now be misaligned with the `text-input` which would be `48px` tall?

photodow commented on packages/components/src/components/button/_mixins.scss:
```diff
@@ -42,8 +296,9 @@
     position: absolute;
     right: rem(16px);
```
So the icon only button looks great since it's sitting center in a `48x48` box. It's my understanding the other icons are supposed to take on this same relationship, but as the icon grows it shifts to the `left` and becomes off-center since it's absolutely positioned `right`. This is a common problem across many of the components. I believe I recall seeing a couple of scenarios where the icon needs to be vertically adjusted because of the same problem.

I'm seeing this in all of the example components you have in this RFC so far.

<img src="https://user-images.githubusercontent.com/3793636/62341351-96dbbf80-b4a8-11e9-9c98-448db9fbaa19.png" width="220px" />

mattrosno replied:
I defer to design on this. Buttons with an icon, the icon is positioned 16px from the right. I didn't change that. Icon-only buttons, that right spacing is 14px computed. The default button with icon can be refactored to make those align, if that's the intent. It's good to make sure that this approach can accomplish that, which it can. IMO making the button component perfect is outside the scope of the RFC.

photodow replied:
@mattrosno please don't think I'm asking you to make this RFC perfect. I don't expect you to correct this in the RFC necessarily. I'm only trying to point out moments where it doesn't align with our needs to help validate whether this is the correct approach. If something is missing I don't want people to think this is the perfect solution when it's still missing a couple of scenarios.

I'll see if @oliviaflory can chime in on this one too.

oliviaflory replied:
Yes, the button with icon should be positioned 14px from the right in order to achieve the centering within `48x48` box. This will apply to all components with icons positioned on the right.

<img width="596" alt="Screen Shot 2019-08-02 at 3 23 38 PM" src="https://user-images.githubusercontent.com/45692622/62393984-dbfc0200-b539-11e9-9490-d1ef09bcc30e.png">

photodow commented on packages/components/src/components/text-input/_mixins.scss:
```diff
@@ -0,0 +1,150 @@
+      right: rem(16px);
+      fill: $support-01;
+      height: $text-input-icon-size;
+      width: $text-input-icon-size;
```
So this looks like its working at the moment for the invalid icon, but I wanted to point out that the password show/hide icon does not get resized with this update. It's still `16px` instead of the desired `20px` since the font increased.

<img src="https://user-images.githubusercontent.com/3793636/62342170-837e2380-b4ab-11e9-9101-8c9077ce3eb1.png" width="250px" />

mattrosno replied:
Good catch. I didn't hit every icon selector in text-input component. It just needs to use the same variable as the invalid icon:

```scss
.#{$prefix}--text-input--password__visibility {
  // styles

  svg {
    height: $text-input-icon-size;
    width: $text-input-icon-size;
    // styles
  }
}
```

photodow replied:
Maybe? It might not be that simple here? When I tried `height` and `width` last night it didn't work. Maybe something wrong the SVG's `preserveAspectRatio`? Just guessing at this point since I haven't had time to figure it out yet.

photodow commented on packages/components/src/components/text-input/_mixins.scss:
```diff
@@ -0,0 +1,150 @@
+
+    .#{$prefix}--text-input--invalid
+      + .#{$prefix}--text-input--password__visibility {
+      right: rem(40px);
```
According to the specs something else we need to consider is the spacing of a row of icons. Based on the specs it's my understanding they need to have `mini-unit(1)` in-between them. As an example, the invalid password's show/hide icon is crowding the invalid icon.

<img src="https://user-images.githubusercontent.com/3793636/62342591-05228100-b4ad-11e9-82d8-ceb02ce043c8.png" width="125px" />

That adjustment may also create a cascading effect where the text-input `padding-right also needs to be updated to reflect the new icon and spacing adjustments.

<img src="https://user-images.githubusercontent.com/3793636/62342704-74987080-b4ad-11e9-9e4f-04087a5b5776.png" width="125px" />

mattrosno replied:
We could calculate those right positionings relative to the icon size, that we have in a Sass variable. I didn't take the RFC that far, but it's possible.

photodow commented on packages/react/.storybook/_container.scss:
```diff
@@ -1,9 +1,108 @@
+@import '~carbon-components/src/components/button/mixins';
+@include carbon--type-theme($button-type-tokens) {
+  @include exports('button') {
+    @include button;
```
One of the things about this RFC that makes me feel uncomfortable is that we're opening up the door for zoning. I'm struggling to think of a use case that can't be solved in a better way. It just feels like this is opening up a can of worms more than it's solving a problem. 

<img src="https://user-images.githubusercontent.com/3793636/62346097-77e62900-b4ba-11e9-94c7-fb7988a06aa2.png" width="270px" />

I wonder if there is a way to solve this with a global switch or import like we've talked about in the past where it's an all or nothing approach.

```scss
@import '@carbon/dotcom-theme';
```

```scss
$type-theme: 'expressive';
```

photodow replied:
One last thing to consider about this RFC is that it feels like we might be unintentionally discouraging a unified design system. Instead of going to carbon for the one component, our adopters will now have to decide between two components y'alls and ours. Instead of unifying it's almost like we're continuing to divide. Just something to think about and consider as we decide to move forward.

```
@carbon/accordion
@carbon/dotcom-accordion

@carbon/button
@carbon/dotcom-button

@carbon/text-input
@carbon/dotcom-text-input
```

photodow reviewed (commented):
@mattrosno so I commented on a couple of things that it looks like you missed. Sorry if it's too detailed for this discussion. I realize this isn't a final proposal, or even trying to be merged at this point. I just want to make sure some of these common edge cases don't get overlooked as we push forward.

I also documented a couple of my concerns we've talked about together for others to consider as we look at this RFC.

Thanks again so much for the work on this. I'm absolutely loving some of the padding and icon calculations you're doing based on the type tokens.

mattrosno commented:
Closing this RFC in favor of a new theming RFC coming soon!
